patroni/postgresql/__init__.py

```diff
@@ -1093,29 +1106,28 @@ def pick_synchronous_standby(self, cluster, sync_node_count=1, sync_node_maxlag=
         """
         if self._major_version < 90600:
             sync_node_count = 1
-        members = {m.name.lower(): m for m in cluster.members}
+        members = CaseInsensitiveDict({m.name: m for m in cluster.members})
         candidates = []
         sync_nodes = []
         replica_list = []
         # Pick candidates based on who has higher replay/remote_write/flush lsn.
-        sync_commit_par = self._get_synchronous_commit_param()
-        sort_col = {'remote_apply': 'replay', 'remote_write': 'write'}.get(sync_commit_par, 'flush')
+        synchronous_commit = self._cluster_info_state_get('synchronous_commit')
+        sort_col = {'remote_apply': 'replay', 'remote_write': 'write'}.get(synchronous_commit, 'flush') + '_lsn'
+        pg_stat_replication = [(r['application_name'], r['sync_state'], r[sort_col])
+                               for r in self._cluster_info_state_get('pg_stat_replication') or []
+                               if r[sort_col] is not None]
         # pg_stat_replication.sync_state has 4 possible states - async, potential, quorum, sync.
         # Sort clause "ORDER BY sync_state DESC" is to get the result in required order and to keep
         # the result consistent in case if a synchronous standby member is slowed down OR async node
         # receiving changes faster than the sync member (very rare but possible). Such cases would
         # trigger sync standby member swapping frequently and the sort on sync_state desc should
         # help in keeping the query result consistent.
-        for app_name, sync_state, replica_lsn in self.query(
-                "SELECT pg_catalog.lower(application_name), sync_state, pg_{2}_{1}_diff({0}_{1}, '0/0')::bigint"
-                " FROM pg_catalog.pg_stat_replication"
-                " WHERE state = 'streaming' AND {0}_{1} IS NOT NULL"
-                " ORDER BY sync_state DESC, {0}_{1} DESC".format(sort_col, self.lsn_name, self.wal_name)):
+        for app_name, sync_state, replica_lsn in sorted(pg_stat_replication, key=lambda r: (r[1], r[2]), reverse=True):
             member = members.get(app_name)
-            if member and not member.tags.get('nosync', False):
+            if member and member.is_running and not member.tags.get('nosync', False):
                 replica_list.append((member.name, sync_state, replica_lsn, bool(member.nofailover)))
 
-        max_lsn = max(replica_list, key=lambda x: x[2])[2] if len(replica_list) > 1 else int(str(self.last_operation()))
```

Comment:
can't get it - what has changed so that we don't need `int(str())` anymore?

Reply:
The last_operation() method used to return str. It was changed in https://github.com/zalando/patroni/commit/935805087f8805a4c9dad3401311e05972ef4f20#diff-72030eab48ce77175c998a2fd1ee9e84fe950e132b66972aa72496532a3923e5R876-R878